Superplot widget

gui-co opened the pull request:
**Description of work.**

This PR provides a new widget to help users in over-plotting data. Like the "fit browser", Superplot is a decorator widget of the plot window.

**To test:**

* code review
* functional tests: plot any data and manipulate the Superplot widget
* integration with the current plot related functionalities (i.e. saving of plots in the project, curve fitting, ...)

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Commit:
[SPLOT] change button text in plot dialog

Commit:
[SPLOT] a spinbox to change the current workspace

With this commit, the current workspace can be changed either with a
slider or with a spinbox. The value of the two widgets is synchronised.

Commit:
[SPLOT] refactor update plot function

Commit:
[SPLOT] spinbox to select the current spectrum

Commit:
[SPLOT] remove plot is ws is removed

Commit:
[SPLOT] add all members of a workspace group

Commit:
[SPLOT] update current workspace with selection

With this commit, the current workspace can be selected by highlighting
it in the workspace list. It will update the sliders positions and the
plot.

Commit:
[SPLOT] instanciate on each opening

Commit:
[SPLOT] instantiate through prensenter

Commit:
[SPLOT] remove useless functions

Commit:
[SPLOT] initial state of the superplot

When the superplot is started, the widget state is synchronised with the
currently plotted data.

Commit:
[SPLOT] memorised the current spectrum

Commit:
[SPLOT] refactoring in presenter

Commit:
[SPLOT] block signals of slider

Commit:
[SPLOT] typo

Commit:
[SPLOT] refactoring in presenter

Commit:
[SPLOT] sliders and spinboxes start at 1

Commit:
[SPLOT] avoid replotting everything

This commit introduces a new member variable that keeps track of the
last modified plotted workspace, spectrum index. When the plot is
updated, only this last modified curve needs to be removed (if the hold
button is not active on it).

Commit:
[SPLOT] reconnect the close function

Commit:
[SPLOT] select last added workspace

Commit:
[SPLOT] use axesfunctions.plot

Commit:
[SPLOT] use wkspIndex instead of specNum

Commit:
[SPLOT] refactoring

Commit:
[SPLOT] update plot to properly display it

To avoid that the axis stay behind the dockwidgets, the
`onVisibilityChanged` signal of one of the dockwidget is now connected
to an update of the figure.

Commit:
[SPLOT] split toggle function in add and remove

Commit:
[SPLOT] bug in last workspace selection

Commit:
[SPLOT] bins and spectra modes

This commit adds the support for 'bins' and 'spectra' modes.

Commit:
[SPLOT] avoid recursive calls

Commit:
[SPLOT] remove workspace slider and spinbox

Commit:
[SPLOT] change from QList to QTree

Commit:
[SPLOT] add sublist of spectra

Commit:
[SPLOT] hold button adds spectrum to the list

Commit:
[SPLOT] multiple selection in workspace tree

Commit:
[SPLOT] hold multiple at the same time

Commit:
[SPLOT] remove useless call

Commit:
[SPLOT] update ws tree when adding/removing

Commit:
[SPLOT] modify plot behavior

When workspace(s) are selected, the first spectrum (bin) is plotted.

Commit:
[SPLOT] adapt selection method

Commit:
[SPLOT] method to add and remove single workspace

This commit adds a method to add and a method to remove a single
workspace to/from the view. This avoid a complete reset of the list.

Commit:
[SPLOT] allow removing of several workspaces

Commit:
[SPLOT] update spectrum slider with selection

When the selection changes, the spectrum slider represents:
* the selected spectrum if the selection is homogenous
* 0 if workspaces are selected or the selection is not homogenous

Commit:
[SPLOT] select first spectrum when un-hold

Commit:
[SPLOT] bug on plot update

Commit:
[SPLOT] new classes to represent tree items

Commit:
[SPLOT] new function to modify labels in tree view

Commit:
[SPLOT] update tree view with labels and colors

With this commit, the workspace tree is updated according to the plotted
data. The spectra are colored according to the corresponding curve and
titles match the matplotlib legend.

Commit:
[SPLOT] delete button for each workspace

Commit:
[SPLOT] typo in member name

Commit:
[SPLOT] delete button for each spectrum

This commits adds a delete button in front of each selected spectrum in
the list. This button allow the user to deselect the correcponding
spectrum.

Commit:
[SPLOT] remove main delete button

Commit:
[SPLOT] add column titles

Commit:
[SPLOT] spectrum/bin combobox on the left side

Commit:
[SPLOT] link spectrum to workspace

When the delete button associated to a spectrum is pressed, the signal
contains the spectrum index but also the workpace name. This is
necessary to delete the correct spectrum from the list.

Commit:
[SPLOT] connect the individual delete buttons

Commit:
[SPLOT] missing axis type when plotting

Commit:
[SPLOT] avoid selection in empty list

Commit:
[SPLOT] avoid some matplotlib calls if plot empty

Commit:
Revert "[SPLOT] method to add and remove single workspace"

This reverts commit fe0d0bcd2895ab43da45849ccbe9bb7e0cd2ed08.

Commit:
[SPLOT] keep selection when adding ws

Commit:
[SPLOT] clear MantidAxes internal list

This commit adds an override of the axes.clear() method to also clear
the internal MantidAxes list.

Commit:
[SPLOT] remove the dockwidgets when closing

Commit:
[SPLOT] close superplot when plot window is closed

Commit:
[SPLOT] add an ads observer in superplot model

This commit adds an ADS observer. This observer will report any
workspace deletion/renaming to the superplot.

Commit:
[SPLOT] connect ADS observer deletion signal

This commit adds the connections with the ADS observer workspace
deletion signal. This signals is propagated to the view, the workspace
is deleted from the list.

Commit:
[SPLOT] connect ADS workspace rename signal

This commit connects the ADS observer signal that corresponds to a
workspace renaming.

Commit:
[SPLOT] reset plot layout when resizing

Commit:
[SPLOT] connect ADS replace workspace signal

This commit connect the signal that is triggered when a workspace is
replaced in the ADS.

Commit:
[SPLOT] slider active when spectrum selected

The slider was not active when a spectrum was selected in the list. With
this commit, one can slide over spectra starting from the selected one.

Commit:
[SPLOT] move ui files

Commit:
[SPLOT] expand tree widget when ws is added

Commit:
[SPLOT] bug is slider max value

Commit:
[SPLOT] uncheck button after hold

Commit:
[SPLOT] remove selection when unhold

Commit:
[SPLOT] clear selection when deleting ws/sp

Commit:
[SPLOT] update slider when selection changed

Commit:
[SPLOT] disable hold button when no selection

Commit:
[SPLOT] bug, set max before position

Commit:
[SPLOT] setSelection method

Commit:
[SPLOT] don't change selection after hold

Commit:
[SPLOT] block signals when setting the tree widget

Commit:
[SPLOT] adjust column size to centents

Commit:
[SPLOT] fix workspace list minimum size

Commit:
[SPLOT] initial state

With this commit, the initial state of the superplot widget is
synchronized with the current plot.

Commit:
[SPLOT] expand tree widget when sp is/are added

Commit:
[SPLOT] deal with multiple selection

When multiple spectra are selected, the slider is disabled.

Commit:
[SPLOT] spectrum slider depends on mode

Commit:
[SPLOT] refactoring

Commit:
[SPLOT] increase size of column

Commit:
[SPLOT] keep axes scale when plotting

Commit:
[SPLOT] update mode when deleting last ws

Commit:
[SPLOT] avoid hard crash when resizing failed

Commit:
[SPLOT] avoid adding 2 times same ws from group

This commit adds a check on every workspaces of a group to be sure that
they are not already present in the model.

Commit:
[SPLOT] spectra doct is useless now

This commits removes the spectra dictionnary from the model. This is not
used anymore, the spectra position is given either by the selected
workspace spectra pair, or by the spectrum slider position.

Commit:
[SPLOT] add constant values for model plot modes

Commit:
[SPLOT] refactoring

Commit:
[SPLOT] synchronise model and view at init

Commit:
[SPLOT] avoid overplotting same data

Commit:
[SPLOT] check data exist before removing

Commit:
[SPLOT] argument not used

Commit:
[SPLOT] reset mode when last plotted ws is deleted

Commit:
[SPLOT] no signal when workspace not in model

Commit:
[SPLOT] add model unit tests

Commit:
[SPLOT] presenter unit test

Commit:
[SPLOT] used matplotlib gca function

Commit:
[SPLOT] plot current spectrum if not

Commit:
[SPLOT] update list creation

Commit:
[SPLOT] refactoring

Commit:
[SPLOT] use getSelection in onWorkspaceRenamed

Commit:
[SPLOT] use getSelection in onWorkspaceDeleted

Commit:
[SPLOT] refactoring

Commit:
[SPLOT] use getSelection in onModeChanged

Commit:
[SPLOT] update slider based on selection

Commit:
[SPLOT] refactoring

Commit:
[SPLOT] refactoring

Commit:
[SPLOT] coherent selection after spectrum deletion

Commit:
[SPLOT] keep plot param when reploting

With this commit, the plot keeps all the axes parameters (scale, limits,
...) when the plot is reloaded. This ensures that the zoom level, the
axes scales or any plot parameters are not erased when the superplot is
used.

Commit:
[SPLOT] hold button update based on seelction

Commit:
[SPLOT] update spectrum slider based on selection

Commit:
[SPLOT] reset legend of matplotlib

Commit:
[SPLOT] plot current spectrum only if enabled

Commit:
[SPLOT] uncheck hold when no selection

Commit:
[SPLOT] split hold button slot in two

Commit:
[SPLOT] remove margins

Commit:
[SPLOT] do not disable hold button

Commit:
[SPLOT] round 180 to 200

Commit:
[SPLOT] change used qtreewidget signal

Commit:
[SPLOT] missing bin mode

Commit:
[SPLOT] remove unused function

Commit:
[SPLOT] use getSelection

Commit:
[SPLOT] improve selection

Add a special value in selection to be able to select the workspace
name.

Commit:
[SPLOT] remove useless call

Commit:
[SPLOT] remove not used functions

Commit:
[SPLOT] better method name

Commit:
[SPLOT] view unit test

Commit:
[SPLOT] superplot in context menu

Commit:
[SPLOT] update plot with drag and drop

Commit:
Revert "[SPLOT] change used qtreewidget signal"

This reverts commit 6edf346df588f53f0dc7913b1ecffb3f8f623cba.

Commit:
[SPLOT] new method to remove all curves

Unlike clear(), remove_all_curves() keeps the axes limits, the axes
scale, ...

Commit:
[PLOT] save plot axis type

Commit:
[SPLOT] select mode from current plot at init

Commit:
[SPLOT] use specNum instead of wkspIndex

When saving/recovering the project, the specNum is used. This change is
needed to properly recover curves plotted through the superplot.

Commit:
[SPLOT] adapt presenter tests

Commit:
[SPLOT] update plot with drag and drop

Commit:
[SPLOT] avoid update of hidden superplot

Commit:
[SPLOT] rescale when changing mode

Commit:
[SPLOT] new function to synchronize plot

Commit:
[SPLOT] refactoring

Commit:
[SPLOT] refactoring of plot update

Commit:
[SPLOT] connect context menu

Commit:
[SPLOT] fit and superplot mutually exclusive

Commit:
[SPLOT] bug - get the correct wkspIndex for bins

Commit:
[SPLOT] get wkspIndex from creation_args

When saving/reloading bin curves, the wkspIndex cannot be recovered by
get_artists_workspace_and_workspace_index().

Commit:
[SPLOT] documentation

Harrietbrown commented:
when maximizing the superplot window it does not resize the plot.
1. superplot a workspace with multiple spectra
2. resize the window so that the plot is small
3. maximize the window
4. the plot remains small in the maximized window
5. if you remove a spectra it resizes to what it should be.

Harrietbrown commented:
What is the functional difference between plotting with the plot option on the context menu and the superplot option?

Harrietbrown commented:
Once you have the superplot open you can add table workspaces to the list. this raises an error as it has no histograms.

Harrietbrown commented:
On project recovery if the plot has two different workspaces I get this error when trying to open superplot on that plot.

```
Traceback (most recent call last):
  File "c:\users\wey38795\documents\mantid_builds\master\raw\mantid\qt\applications\workbench\workbench\plotting\figuremanager.py", line 403, in superplot_toggle
    self.superplot = Superplot(self.canvas, self.window)
  File "c:\users\wey38795\documents\mantid_builds\master\raw\mantid\qt\python\mantidqt\widgets\superplot\SuperplotPresenter.py", line 42, in __init__
    self._updatePlot()
  File "c:\users\wey38795\documents\mantid_builds\master\raw\mantid\qt\python\mantidqt\widgets\superplot\SuperplotPresenter.py", line 267, in _updatePlot
    specNum = ws.getSpectrumNumbers()[sp]
IndexError: list index out of range
```

Harrietbrown commented:
When you turn superplot on it turns off error bars.

When you overplot a spectrum from the workspace widget it does not update the superplot tabs.

Harrietbrown commented on qt/python/mantidqt/widgets/superplot/test/SuperplotModelTest.py:
```diff
@@ -0,0 +1,202 @@
+        wsList = self.model.getWorkspaces()
+        self.assertEqual(wsList, self.model._workspaces)
+        wsList = wsList[0:2]
+        self.assertEqual(wsList, ["ws1", "ws2"])
```
could you just add a comment that this is to make sure that getWorkspace is not returing a pointer to its own list?

gui-co replied:
done

Harrietbrown reviewed (changes requested):
This has been some great work so far, it will be a great addition to our plotting when finished.

Commit:
[SPLOT] move resize event to window level

Commit:
Export workspace selector method in python

Commit:
[SPLOT] restrict combobox to supported workspaces

Commit:
[SPLOT] use errorbar plotting method when needed

Commit:
[SPLOT] add eventworkspace in combobox

Commit:
[SPLOT] typos

Commit:
[SPLOT] document workspace addition

Commit:
[SPLOT] remove useless instruction

Commit:
[SPLOT] list shallow copy

Commit:
[SPLOT] document test

Commit:
[SPLOT] connect superplot context menu

Commit:
[SPLOT] update test for superplot

gui-co commented:
Thanks for the review and comments.

I answered most of your points. I cannot reproduce your fourth comment ("On project recovery if the plot has two different workspaces I get this error when trying to open superplot on that plot."). It appears when you save and load the project ?

For the rest:
* I moved the update on resize to the plot window itself. The plot should now be resized when the window is resized and/or maximized.
* The context menu was not properly connected. It should now open the plot with the superplot widget on.
* I filtered out table workspaces from the combobox.
* I resolved the problem with erro bars

gui-co commented on docs/source/workbench/superplot.rst:
```diff
@@ -0,0 +1,40 @@
+Superplot is a decorator widget of the plot window. It facilitates over-plotting
+and manipulation of overplotted data. The superplot is accessible:
+
+* directly in the plot window through the menu bar
```
done

gui-co commented on docs/source/workbench/superplot.rst:
```diff
@@ -0,0 +1,40 @@
+and manipulation of overplotted data. The superplot is accessible:
+
+* directly in the plot window through the menu bar
+* in the context menu of supported workspaces
```
done

gui-co commented on docs/source/workbench/superplot.rst:
```diff
@@ -0,0 +1,40 @@
+
+The superplot widget is composed of two distinct parts:
+
+* left side of the plot window: a list of selected workspaces and spectra. This
```
done

gui-co commented on docs/source/workbench/superplot.rst:
```diff
@@ -0,0 +1,40 @@
+
+* left side of the plot window: a list of selected workspaces and spectra. This
+  list contains workspaces names and under them the plotted data if any.
+* bottom of the plot: a navigation bar that contains a slider and different
```
done

gui-co commented on docs/source/workbench/superplot.rst:
```diff
@@ -0,0 +1,40 @@
+   :align: center
+
+
+**Usage**
```
done

gui-co commented on qt/python/mantidqt/widgets/superplot/SuperplotModel.py:
```diff
@@ -0,0 +1,223 @@
+        super().__init__()
+        self._workspaces = list()
+        self._plottedData = list()
+        self._plotMode = None
```
I removed that line.

gui-co commented on qt/python/mantidqt/widgets/superplot/SuperplotModel.py:
```diff
@@ -0,0 +1,223 @@
+
+    def __init__(self):
+        super().__init__()
+        self._workspaces = list()
```
I can't as it would be a class variable. All instances of superplot would modify the same list.

gui-co commented on qt/python/mantidqt/widgets/superplot/SuperplotModel.py:
```diff
@@ -0,0 +1,223 @@
+        Returns:
+            list(str): list of workspace names
+        """
+        return [name for name in self._workspaces]
```
done

gui-co commented on qt/python/mantidqt/widgets/superplot/SuperplotModel.py:
```diff
@@ -0,0 +1,223 @@
+        Returns:
+            list(tuple(str, int)): list of workspace, spectrum index pairs
+        """
+        return [data for data in self._plottedData]
```
done

gui-co commented on qt/python/mantidqt/widgets/superplot/SuperplotPresenter.py:
```diff
@@ -0,0 +1,488 @@
+
+class SuperplotPresenter:
+
+    BIN_MODE_TEXT = "Bin"
```
They do not represent the same concept. In the presenter they are strings for the combobox, in the model they replace an enum.

gui-co commented on qt/python/mantidqt/widgets/superplot/SuperplotPresenter.py:
```diff
@@ -0,0 +1,488 @@
+        """
+        selection = self._view.getSelection()
+        if wsName is None:
+            selectedWorkspaces = [ws for ws in selection]
```
done

gui-co commented on qt/python/mantidqt/widgets/superplot/SuperplotPresenter.py:
```diff
@@ -0,0 +1,488 @@
+                args = axes.creation_args
+            except:
+                args = [{}]
+            if "axis" in args[0]:
```
You mean the search for the axis key? Here I am checking that all curves are either using `MantidAxType.BIN` or `MantidAxType.SPECTRUM` axis. From what I understood of these `creation_args` is that if axis is not present, `MantidAxType.SPECTRUM` is used. Moreover, the axis type is a python enum and sometime the value is used instead of the name (`MantidAxType.SPECTRUM.value`, `MantidAxType.BIN.value`).

Harrietbrown commented:
This project has the files that cause the error.
[31375_project.zip](https://github.com/mantidproject/mantid/files/6502419/31375_project.zip)

Commit:
Merge branch 'master' into superplot

Commit:
[SPLOT] avoid crash on resize

When the plot was resized to a size that is too small to fit the title,
legend, axis... it was crashing.

Commit:
[SPLOT] bug in axis type

Commit:
[SPLOT] remove useless instructions

Commit:
[SPLOT] do not replot every curves

Commit:
[SPLOT] update presenter tests

Harrietbrown reviewed (dismissed):
Well done, you've fixed the issues I've found and the superplot looks like it is working really well.
Good job

gui-co commented:
Thanks for the review.

I did not solve the problem with your file but investigated a bit.
To get the currently plotted data, I am using `get_artists_workspace_and_workspace_index()`:
https://github.com/mantidproject/mantid/blob/ce3cfe9247f560c2ec6542612ed3eca05b3f2e75/Framework/PythonInterface/mantid/plots/mantidaxes.py#L211

With the plot in your project, this function returns 1 for the workspace index (while the workspace contains only 1 spectrum). If I re-plot the same data, the function returns 0.

Harrietbrown commented:
Strange, because I tried recreating the issue after your latest commits and didn't get it.

gvardany commented:
There are a couple of fixes to come as discussed privately.

Commit:
[SPLOT] avoid crash when adding ws not in ADS

Commit:
[SPLOT] valid state for superplot toolbutton

This commit ensures that the superplot button is in the right state.
This will avoid desynchronization between superplot state and button
state when using the context menu to open the superplot.

Commit:
[SPLOT] keep homogeneity in context menu entries

Commit:
[SPLOT] superplot with errors in context menu

Commit:
[SPLOT] give focus to slider container widget

Commit:
[SPLOT] forward some useful key press events

Commit:
[SPLOT] accept focus with click

Commit:
[SPLOT] legend draggable

Commit:
[SPLOT] remove legend and title on empty plot

Commit:
[SPLOT] make a seletion at startup

Commit:
[SPLOT] adapt "unfold" section size

Commit:
[SPLOT] get minimum size from contents

Commit:
[SPLOT] remove axes when plot is empty

Commit:
[SPLOT] tooltips

gui-co commented:
Here is the list of changes:
* error caught when adding a workspace that does not exist
* synchronization between button state and widget status when superplot is opened through the context menu
* `Superplot` changed to `Superplot...` in context menu
* `Superplot with errors...` added in context menu
* in the QListWidget, the `unhold` column is smaller and has a fixed size
* the last spectrum of each workspace is pre-selected when the superplot is started
* space key bound to the `hold` action and arrow keys to the slider movements
* bottom QDockWidget has focus when the superplot is started
* labels, title and legend are removed when the plot is empty
* tootlips have been added for the different widgets
* plot legend is draggable

Commit:
[SPLOT] test non existing workspace

gvardany commented:
Could you also add a release note, you can put the image from the docs there as well.

gvardany commented:
I'm getting an error now on macOS, is this because of the new changes?
![Screenshot 2021-06-03 at 10 19 05](https://user-images.githubusercontent.com/16865759/120612162-32e40280-c455-11eb-8e92-3607d58fba10.png)

gvardany commented:
As I mentioned, we need to increase the default plot window size. In principle, this can be a user setting, but the current default is too small.
As you can see, the Superplot button is not visible on mac, unless you click on the arrow to expand the tool buttons.
![Screenshot 2021-06-03 at 10 22 32](https://user-images.githubusercontent.com/16865759/120612637-b0a80e00-c455-11eb-8e10-24cc0fffeb8a.png)

Commit:
[SPLOT] release note

Commit:
[SPLOT] Legend.draggable deprecated since mpl 3

gvardany commented:
When I shrink the window and then stretch it bigger again, the slider doesn't fill in the original space. 
![Screenshot 2021-06-03 at 11 07 08](https://user-images.githubusercontent.com/16865759/120618992-dfc17e00-c45b-11eb-8359-52bbb3c3661a.png)

gvardany commented:
When you plot from the context menu, there is an option in the dialog to chose between Individual, Tiled and Waterfall.
As you can see Tiled doesn't make any sense for superplot, so should be disabled.
Waterfall does work, it's just that the Superplot button goes missing from the toolbar.
Not sure if Fit button is supposed to be there too.

![Screenshot 2021-06-03 at 11 15 06](https://user-images.githubusercontent.com/16865759/120620210-08964300-c45d-11eb-953a-fc2d88bc0bd7.png)

![Screenshot 2021-06-03 at 11 12 57](https://user-images.githubusercontent.com/16865759/120620205-07651600-c45d-11eb-9309-4feecefad56f.png)

gvardany commented:
When I start superplot on a sample workspace sometimes the curves appear in strange order in the list:
This is for the indices 7-9, for example:
![Screenshot 2021-06-03 at 11 21 49](https://user-images.githubusercontent.com/16865759/120621323-11d3df80-c45e-11eb-8c7f-85669e901190.png)

gvardany commented on docs/source/workbench/superplot.rst:
```diff
@@ -0,0 +1,41 @@
+Superplot is a decorator widget of the plot window. It facilitates over-plotting
+and manipulation of overplotted data. The superplot is accessible:
+
+* Directly in the plot window through the menu bar
```
It's not the menu bar, it's the toolbar.

gui-co replied:
done

gvardany commented on docs/source/workbench/superplot.rst:
```diff
@@ -0,0 +1,41 @@
+and manipulation of overplotted data. The superplot is accessible:
+
+* Directly in the plot window through the menu bar
+* In the context menu of supported workspaces
```
Under the submenu Plot.

gui-co replied:
done

gvardany commented on docs/source/workbench/superplot.rst:
```diff
@@ -0,0 +1,41 @@
+
+Each workspace and bin/spectrum is associated, in the list, with a delete
+button. To remove a curve, one can also slide to the specific position and
+release the hold button.
```
This text misses the main point, which is that the current curve indicated by the slider for the workspaces that are selected from the list is plotted, but not persisted. One can move the slider to iterate through the curves one by one, and, if needed, click Hold to persist a curve on the plot. 
Also, for the raw data, the legend shows the spectrum number, whereas the spinbox in superplot refers to the workspace index, as it should be. The index always starts from 0 and is contiguous, whereas the spectrum number starts from 1 and might not be contiguous if there are grouped detectors. So for raw data the index indicated in the spinbox will often be 1 less than the spectrum number in the legend/workspace list. This has to be highlighted in the docs.

gui-co replied:
done

gvardany reviewed (commented):
a few comments on the documentation:

Commit:
[SPLOT] update documentation

Commit:
[SPLOT] adapt plot window width

To properly display the toolbar (with the superplot button), the plot
window width is calculated from the maximum width between toolbar and
canvas.

Commit:
[SPLOT] no superplot with tiled/waterfall plots

When the superplot is started through the context menu, one can chose
"waterfall" and "tiles" mode. In these cases, the superplot widget makes
no sense and is then not started.

Commit:
Merge branch 'master' into superplot

gui-co commented:
* The documentation is updated
* The superplot is disabled when "waterfall" or "tiles" mode is selected
* The size of the window is calculated from the toolbar (tested on mac)

Concerning the order of curves, it comes from the function  `get_artists_workspace_and_workspace_index()` and correspond to the order in the legend of the original plot.
https://github.com/mantidproject/mantid/blob/ce3cfe9247f560c2ec6542612ed3eca05b3f2e75/Framework/PythonInterface/mantid/plots/mantidaxes.py#L211
I can reorder them but i will also need to re-plot everything to stay coherent with the legend.

Concerning the size of the slider, I cannot reproduce it on my mac VM.

gvardany commented on qt/applications/workbench/workbench/plotting/figuremanager.py:
```diff
@@ -237,15 +240,24 @@ def __init__(self, canvas, num):
         height = cs.height() + self._status_and_tool_height
-        self.window.resize(cs.width(), height)
+        # get window width from toolbar/canvas width
+        if cs.width() > ts.width():
```
This works, but I wonder if one should adjust also the height to keep the aspect ratio the same.
```suggestion
 width = max(cs.width(), ts.width())
```

gui-co replied:
I changed the if else statement with max(). For the aspect ratio, i do not understand which one. The window height comes from the canvas + the toolbar, and the width from the max between the two widgets. Which ratio should be kept constant?

gvardany replied:
I think the canvas aspect ratio should be kept constant for the time being.
Still it is dangerous to take the max of the toolbar as it can potentially grow further.
So the best would be to make them a user setting, so one can customize also based on the screen size and resolution. I've opened another issue for this: https://github.com/mantidproject/mantid/issues/31724

gvardany replied:
This is what I get now. There is a fair bit of empty space in the toolbar, even though we take its size. Any idea?
![Screenshot 2021-06-07 at 10 05 21](https://user-images.githubusercontent.com/16865759/120981395-eca1e280-c777-11eb-8d2b-96c6ac22efad.png)

Commit:
[SPLOT] use max() instead of if else

gvardany commented:
I've opened this https://github.com/mantidproject/mantid/issues/31723 for the curve order. The issue is there already for regular plots, so should rather be corrected there at source. Could you also create the issue with saving of the project when there are removed curves? This is also not related to superplot itself.

gui-co commented:
The issue is already opened: #31604

Commit:
Revert "[SPLOT] adapt plot window width"

This reverts commit dd9f497e8dbfe91cb5713688ab4e62fc0f6b0774.

gvardany commented:
For the slider on macOS, it resizes correctly with qt 5.14 (and 5.15) which is where the build servers are at. The issue is observed only with 5.13, so presumably it was a bug in qt that got solved. @Harrietbrown could you give another spin and reapprove?

gvardany commented:
My flake8 caught some issues, I wonder why it wasn't flagged by the formatting job: 
```
/usr/local/bin/flake8 SuperplotPresenter.py
SuperplotPresenter.py:12:1: F401 'mantid.plots.axesfunctions.plot' imported but unused
SuperplotPresenter.py:92:5: E129 visually indented line with same indent as next logical line
SuperplotPresenter.py:286:5: E129 visually indented line with same indent as next logical line
SuperplotPresenter.py:380:17: F841 local variable 'currentIndex' is assigned to but never used
SuperplotPresenter.py:425:17: E128 continuation line under-indented for visual indent
SuperplotPresenter.py:450:9: F841 local variable 'selection' is assigned to but never used
SuperplotPresenter.py:505:9: F841 local variable 'plottedData' is assigned to but never used
```

Commit:
[SPLOT] flake8 unused import

Commit:
[SPLOT] flake8 indentation

Commit:
[SPLOT] flake8 unused variable

Commit:
[SPLOT] flake8 unused variable

Commit:
[SPLOT] flake8 unused variable

Commit:
Revert "[SPLOT] new method to remove all curves"

This reverts commit 06c7d968b01c41561204b2eb86f9612c80b04fc8.

Commit:
Revert "[SPLOT] clear MantidAxes internal list"

This reverts commit 04367665a5126a675bf683650d6d2cc45ec013e3.

gui-co commented:
I resolved the flake8 warnings. I also reverted two commits that introduced functions that are not used anymore.

gvardany commented:
The formatting job says:
```flake8...................................................................Passed```
Any idea why it wasn't caught? Can you check the excluded directories in the config files?

gvardany commented:
OK, the flake8 configuration issue is found and will be fixed separately. @Harrietbrown waiting for your action now.

thomashampson commented:
I noticed that the new python files don't follow the Mantid Python coding standards (including file names, function and variable names)

gui-co commented:
@thomashampson, where can I find the Mantid Python coding standards? I will change that.

thomashampson commented:
@gui-co They are here:
https://developer.mantidproject.org/Standards/PythonStandards.html
We are basically following the PEP 8 style guides, except for certain circumstances,  e.g. when interacting with Qt objects.

Commit:
[SPLOT] snake_case in SpectrumItem methods

Commit:
[SPLOT] snake_case SpectrumItem variables

Commit:
[SPLOT] snake_case for WorkspaceItem methods

Commit:
[SPLOT] snake_case for WorkspaceItem variables

Commit:
[SPLOT] snake_case for SuperplotAdsObserver

Commit:
[SPLOT] snake_case for view methods

Commit:
[SPLOT] snake_case for view variables

Commit:
[SPLOT] snake_case for model methods

Commit:
[SPLOT] snake_case for model variables

Commit:
[SPLOT] snake_case for presenter methods

Commit:
[SPLOT] snake_case for presenter variables

Commit:
[SPLOT] snake_case in model tests

Commit:
[SPLOT] snake_case for view tests

Commit:
[SPLOT] unused imports

Commit:
[SPLOT] snake_case for presenter tests

gui-co commented:
@thomashampson, I think i changed every names. I did not modify the file names, they currently match the name of the class (the main class if there are several).

Could you take a look and tell me if I should change the file names too?

thomashampson commented:
@gui-co, the PEP 8 naming style for module names is all lower-case , with optional underscores to aid readability ([see here](https://www.python.org/dev/peps/pep-0008/#package-and-module-names)). However, in some parts of the code base this isn't followed (e.g. Framework\PythonInterface\plugins\algorithms). It's best to have a look at what's been done in the area of the code that you've added the files to. For the superplot work you've done within widgets that would be the PEP 8 style.

Commit:
[SPLOT] rename to model.py

Commit:
[SPLOT] rename to presenter.py

Commit:
[SPLOT] rename to view.py

Commit:
[SPLOT] snake_case for ui files

Commit:
[SPLOT] merge view related classes

Commit:
[SPLOT] flake8

Commit:
[SPLOT] snake_case

Commit:
[SPLOT] snake_case for tests

gvardany commented:
@thomashampson We would like to get this in soon. Are you happy with the naming changes? The failing system test is on master and is unrelated.

thomashampson commented:
@gvardany, @gui-co Yes the file name formats are now consistent with the others within mantidqt/widgets. 👍

gvardany commented:
Would you (or the original reviewer) mind approving this then?

thomashampson commented on qt/python/mantidqt/widgets/superplot/presenter.py:
```diff
@@ -0,0 +1,506 @@
+        self._view.set_spectrum_spin_box_max(maximum - 1)
+        self._view.set_spectrum_spin_box_value(position)
+
+    def _update_hold_button(self):
```
This function also needs to be called also immediately after the button is pressed (and the spectrum is held) so that its state is correct (maybe in `_on_hold()`).

Additionally, I think it would be a lot more intuitive to change the behaviour of the button so that the text chages from "add" to "remove", rather than the button being checked or unchecked. It is not immediately obvious that the button's check state changes if a spectrum is "held".

thomashampson replied:
This works now

thomashampson commented on qt/python/mantidqt/widgets/superplot/test/test_superplot_presenter.py:
```diff
@@ -0,0 +1,308 @@
+        self.m_view.get_spectrum_slider_position.return_value = 10
+        self.presenter._on_un_hold()
+        calls = [mock.call("ws1", 10), mock.call("ws2", 10)]
+        self.m_model.remove_data.assert_has_calls(calls)
+        self.presenter._update_list.assert_called_once()
+        self.presenter._update_plot.assert_called_once()
+        self.presenter._update_spectrum_slider.assert_called_once()
+
+    def test_on_hold_button_toggled(self):
+        self.presenter._on_hold = mock.Mock()
+        self.presenter._on_un_hold = mock.Mock()
+        self.presenter.on_hold_button_toggled(True)
+        self.presenter._on_hold.assert_called_once()
+        self.presenter._on_un_hold.assert_not_called
+        self.presenter.on_hold_button_toggled(False)
+        self.presenter._on_un_hold.assert_called_once()
+
+    def test_on_mode_changed(self):
+        self.presenter._update_spectrum_slider = mock.Mock()
+        self.presenter._update_plot = mock.Mock()
+        self.m_view.get_selection.return_value = {"ws1": [1], "ws2": [2]}
+        self.presenter.on_mode_changed("mode")
+        self.presenter._update_spectrum_slider.assert_called_once()
+        self.presenter._update_plot.assert_called_once()
+
+    def test_on_workspace_releted(self):
+        self.presenter._update_list = mock.Mock()
+        self.presenter._update_plot = mock.Mock()
+        self.m_view.get_selection.return_value = {"ws1": [1], "ws2": [2]}
+        self.presenter.on_workspace_deleted("ws1")
+        self.m_view.set_selection.assert_called_once_with({"ws2": [2]})
+        self.presenter._update_list.assert_called_once()
+        self.presenter._update_plot.assert_called_once()
+
+    def test_on_workspace_renamed(self):
+        self.presenter._update_list = mock.Mock()
+        self.presenter._update_plot = mock.Mock()
+        self.m_view.get_selection.return_value = {"ws1": [1], "ws2": [2]}
+        self.presenter.on_workspace_renamed("ws1", "ws3")
+        self.m_view.set_selection.assert_called_once_with({"ws3": [1],
```
The `m_` in `self` variables is not necessary here.

gui-co replied:
`m_` means mock, not member.

thomashampson commented on qt/python/mantidqt/widgets/superplot/test/test_superplot_view.py:
```diff
@@ -0,0 +1,235 @@
+qapp = QApplication(sys.argv)
+
+
+class SuperplotViewTest(unittest.TestCase):
```
No need for `m_` in `self` variables.

thomashampson commented on qt/python/mantidqt/widgets/superplot/view.py:
```diff
@@ -0,0 +1,476 @@
+        self._bottom_view.holdButton.setChecked(state)
+        self._bottom_view.holdButton.blockSignals(False)
+
+    def set_spectrum_disabled(self, state):
```
Consider renaming for ease of understanding
```suggestion
    def set_spectrum_slider_disabled(self, state):
```

gui-co replied:
I changed to `set_spectrum_selection_disabled`

thomashampson commented on qt/python/mantidqt/widgets/superplot/view.py:
```diff
@@ -0,0 +1,476 @@
+        self._bottom_view.spectrumSlider.setDisabled(state)
+        self._bottom_view.spectrumSpinBox.setDisabled(state)
+
+    def is_spectrum_disabled(self):
```
Consider renaming for ease of understanding
```suggestion
    def is_spectrum_slider_disabled(self):
```

gui-co replied:
I changed to `is_spectrum_selection_disabled`

thomashampson commented on qt/python/mantidqt/widgets/superplot/view.py:
```diff
@@ -0,0 +1,476 @@
+            nums.append(ws_item.child(i).get_spectrum_index())
+        return nums
+
+    def check_hold_button(self, state):
```
Suggest renaming for clarity (unless you take up my suggestion to rename the button, in which case this is moot)
```suggestion
    def set_hold_button_check_state(self, state):
```

thomashampson reviewed (changes requested):
This is a great feature addition. I've listed below a few things in addition to the in-line comments:

- It would be more intuitive if "hold" and "unhold" were renamed to "add" and "remove", and the hold button's text changed based on whether the spectrum is already added to the list, rather than checking/unchecking the button.
- When selecting the superplot option from the context menu in the workbench, the tiled and waterfall options are present in the dialog. If either are selected, superplot is not enabled. This is confusing bahaviour. Since tiled/waterfall/superplot are currently mutally exclusive, it would make more sense to have the superplot option in the combo box along with tiled and waterfall, rather than having it as an option on the context menu. The same applies to the superplot with errors. Alternatively, the tiled/waterfall combo box could be hidden, but my original suggestion leads to a simplified interface with less code.
- When dragging a workspace from the workbench main window into the superplot, the currently selected plot in the superplot spectrum slider is added to the list of "held" spectra. This seems like a bug to me.

Commit:
[SPLOT] avoid internal check of matplotlib version

Commit:
[SPLOT] homogeneous naming

Commit:
[SPLOT] typo

Commit:
[SPLOT] typo

Commit:
[SPLOT] PR comment

Commit:
[SPLOT] more descriptive variable name

Commit:
[SPLOT] more descriptive variable names

Commit:
[SPLOT] check state is not a text

The hold button is not any more a checkable button. Intead, it carries a
text that informs the user on its state. Two new methods have been added
to the view to set and get this text.

Commit:
[SPLOT] presentation logic of the new hold button

Commit:
[SPLOT] hold button disabled with spectrum slider

gvardany commented:
Thanks for your feedback, here are I my answers to the general points:

> It would be more intuitive if "hold" and "unhold" were renamed to "add" and "remove", and the hold button's text changed based on whether the spectrum is already added to the list, rather than checking/unchecking the button.

I have no objection against this. It is a good suggestion to change the text on the button. One could perhaps change both the text and the button state/background colour.

> When selecting the superplot option from the context menu in the workbench, the tiled and waterfall options are present in the dialog. If either are selected, superplot is not enabled. This is confusing bahaviour. Since tiled/waterfall/superplot are currently mutally exclusive, it would make more sense to have the superplot option in the combo box along with tiled and waterfall, rather than having it as an option on the context menu. The same applies to the superplot with errors. Alternatively, the tiled/waterfall combo box could be hidden, but my original suggestion leads to a simplified interface with less code.

I had the same thought actually when checking tile/waterfall. Since they are mutually exclusive now with superplot, it can indeed go to that same combo-box instead of the context menu. It's true that it will be quite hidden there, so we will need to make a big advertisement for those users that are waiting for this. Just making it more visible shouldn't be the only reason for putting in the context menu. Related to this, I'd also suggest to have the Superplot button (the one next to Fit in the toolbar) visible also on the 2D colorfill plots. Quite often people would want to plot 2D, and then go through line profiles with superplot.

> To improve the user experience and make everything consistent, the spin box should refer to spectrum number and not workspace index. I appreciate that this will be difficult for non-contiguous spectrum number ranges, but in that case it will be even more difficult for the end-user to translate between the two indices.

I have to disagree with this. The slider must be in workspace index. It would be quite strange if it is not contiguous. Even more, I would insist on having the workspace index in the legends too, as it was originally. It was changed to the spectrum number later on without much of discussion/consensus.

thomashampson commented:
> > To improve the user experience and make everything consistent, the spin box should refer to spectrum number and not workspace index. I appreciate that this will be difficult for non-contiguous spectrum number ranges, but in that case it will be even more difficult for the end-user to translate between the two indices.
> 
> I have to disagree with this. The slider must be in workspace index. It would be quite strange if it is not contiguous. Even more, I would insist on having the workspace index in the legends too, as it was originally. It was changed to the spectrum number later on without much of discussion/consensus.
That is fair, I was not aware that the index in the legend was recently changed. I am happy for the slider index to remain as it is based on your explanation. It would be quite a complex change anyway given that the slider can change between spectrum and bin index.

Commit:
[SPLOT] update view tests for the new hold button

Commit:
[SPLOT] update presenter tests for hold button

Commit:
[SPLOT] more descriptive method names

Commit:
[SPLOT] fix indentation

Commit:
[SPLOT] avoid flake8 warning

gui-co commented on qt/python/mantidqt/widgets/superplot/presenter.py:
```diff
@@ -0,0 +1,506 @@
+                if self._matplotlib_version >= "3.0.0":
+                    legend.set_draggable(True)
+                else:
+                    legend.draggable()
```
Done.

gui-co commented on qt/applications/workbench/workbench/plotting/figuremanager.py:
```diff
@@ -337,6 +345,18 @@ def destroy(self, *args):
             # Gcf can get destroyed before the Gcf.destroy
             # line is run, leading to a useless AttributeError.
 
+    def onResize(self):
```
Done.

gui-co commented on qt/python/mantidqt/widgets/superplot/model.py:
```diff
@@ -0,0 +1,227 @@
+    def set_spectrum_mode(self):
+        """
+        Set the plot mode to 'spectra'.
+        """
```
Done.

gui-co commented on qt/python/mantidqt/widgets/superplot/presenter.py:
```diff
@@ -0,0 +1,506 @@
+                i = artists.index(artist)
+                if i >= len(args):
+                    i = 0
+                specIndex = args[i]["wkspIndex"]
```
Done.

gui-co commented on qt/python/mantidqt/widgets/superplot/presenter.py:
```diff
@@ -0,0 +1,506 @@
+
+        for artist in artists:
+            ws, spec_index = \
+                    axes.get_artists_workspace_and_workspace_index(artist)
```
Done.

gui-co commented on qt/python/mantidqt/widgets/superplot/presenter.py:
```diff
@@ -0,0 +1,506 @@
+            self._model.add_workspace(ws_name)
+            self._model.add_data(ws_name, spec_index)
+
+    def on_visibility_changed(self, state):
```
Done.

gui-co commented on qt/python/mantidqt/widgets/superplot/presenter.py:
```diff
@@ -0,0 +1,506 @@
+                    self._view.set_spectrum_disabled(True)
+                    return
+            if mode == self.SPECTRUM_MODE_TEXT:
+                value = ws.getNumberHistograms()
```
Done.

Commit:
[SPLOT] typo in method name

gui-co commented:
I answer most of your comments. Concerning the "hold" button, I changed it to a standard `QPushButton` and changed the text according to the possible action (add or remove). Is a color necessary here ?

Commit:
Hide only seprator

Commit:
[SPLOT] give a workspace list to the superplot

Instead of synchronizing the superplot with the current plot, one can
now give a list of workspaces to the superplot.

Commit:
[SPLOT] open superplot on an empty plot

Commit:
Revert "[SPLOT] superplot with errors in context menu"

This reverts commit 41e827a73fc6a88b54f4f85fb307705fc3468b54.

Commit:
[SPLOT] open empty superplot with context menu

Commit:
Merge branch 'master' into superplot

Commit:
[SPLOT] disable superplot with waterfall/tiled

Commit:
[SPLOT] fix workspace widget tests

Commit:
[SPLOT] naming consistency

thomashampson reviewed (changes requested):
Just a minor change to the documentation required.

Commit:
[SPLOT] naming consistency in documentation

gui-co commented on docs/source/workbench/superplot.rst:
```diff
@@ -0,0 +1,48 @@
+
+Once persisted, the curve(s) is(are) added to the list on the left side. To
+delete a curve, one can use its delete button from the list or slide to the
+specific position and release the hold button.
```
I modified the doc.
